internal/daemon/api_files.go

```diff
@@ -0,0 +1,627 @@
+		}
+		decoder := json.NewDecoder(req.Body)
+		if err := decoder.Decode(&payload); err != nil {
+			return statusBadRequest("cannot decode request body: %v", err)
+		}
+		switch payload.Action {
+		case "make-dirs":
+			return makeDirs(payload.Dirs)
+		case "remove":
+			return removePaths(payload.Paths)
+		case "write":
+			return statusBadRequest(`must use multipart with "write" action`)
+		default:
+			return statusBadRequest("invalid action %q", payload.Action)
+		}
+	default:
+		return statusBadRequest("invalid media type %q", mediaType)
+	}
+}
+
+// Writing files
+
+type writeFilesItem struct {
+	Path        string `json:"path"`
+	MakeDirs    bool   `json:"make-dirs"`
+	Permissions string `json:"permissions"`
+	UserID      int    `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     int    `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+func writeFiles(body io.Reader, boundary string) Response {
+	// Read metadata part (field name "request").
+	mr := multipart.NewReader(body, boundary)
+	part, err := mr.NextPart()
+	if err != nil {
+		return statusBadRequest("cannot read request metadata: %v", err)
+	}
+	if part.FormName() != "request" {
+		return statusBadRequest(`metadata field name must be "request", not %q`, part.FormName())
+	}
+
+	// Decode metadata about files to write.
+	var payload struct {
+		Action string           `json:"action"`
+		Files  []writeFilesItem `json:"files"`
+	}
+	decoder := json.NewDecoder(part)
+	if err := decoder.Decode(&payload); err != nil {
+		return statusBadRequest("cannot decode request metadata: %v", err)
+	}
+	if payload.Action != "write" {
+		return statusBadRequest(`multipart action must be "write", not %q`, payload.Action)
+	}
+	if len(payload.Files) == 0 {
+		return statusBadRequest("must specify one or more files")
+	}
+	infos := make(map[string]writeFilesItem)
+	for _, file := range payload.Files {
+		infos[file.Path] = file
+	}
+
+	errors := make(map[string]error)
+	for i := 0; ; i++ {
+		part, err = mr.NextPart()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return statusBadRequest("cannot read file part %d: %v", i, err)
+		}
+		if !strings.HasPrefix(part.FormName(), "file:") {
+			return statusBadRequest(`field name must be in format "file:/path", not %q`, part.FormName())
+		}
+		p := part.FormName()[len("file:"):]
+		info, ok := infos[p]
+		if !ok {
+			return statusBadRequest("no metadata for path %q", p)
+		}
+		errors[p] = writeFile(info, part)
+		part.Close()
+	}
+
+	// Build list of results with any errors.
+	result := make([]fileResult, len(payload.Files))
+	var firstErr error
+	for i, file := range payload.Files {
+		err, ok := errors[file.Path]
+		if !ok {
+			// Ensure we wrote all the files in the metadata.
+			err = fmt.Errorf("no file content for path %q", file.Path)
+		}
+		if err != nil && firstErr == nil {
+			firstErr = err
+		}
+		result[i] = fileResult{
+			Path:  file.Path,
+			Error: fileErrorToResult(err),
+		}
+	}
```

Comment:
Should we sync the filesystem here?  Please have a look at the atomic write commit we have somewhere in osutil, as this is surprisingly boring.

Reply:
Ha, nice -- osutil.AtomicFile takes care of a lot of this, including setting user and group. I've updated it.